pinot-core/src/main/java/org/apache/pinot/core/indexsegment/mutable/MutableSegmentImpl.java

```diff
@@ -442,13 +470,20 @@ private ColumnDataSource getVirtualDataSource(String column) {
     return null;
   }
 
-  @Override
+  /**
+   * Returns a record that contains only physical columns
+   * @param docId document ID
+   * @param reuse a GenericRow object that will be re-used if provided. Otherwise, this method will allocate a new one
+   * @return Generic row with physical columns of the specified row.
+   */
   public GenericRow getRecord(int docId, GenericRow reuse) {
     for (FieldSpec fieldSpec : _schema.getAllFieldSpecs()) {
       String column = fieldSpec.getName();
-      reuse.putField(column, IndexSegmentUtils
-          .getValue(docId, fieldSpec, _indexReaderWriterMap.get(column), _dictionaryMap.get(column),
-              _maxNumValuesMap.getOrDefault(column, 0)));
+      if (!_schema.isVirtualColumn(column)) {
```

Comment:
Might be better to cache the physical fieldSpecs as we call getRecord() on per record level.

Reply:
done

Reply:
You forgot this one, the most critical one lol

Reply:
Actually, least critical since this is called only during segment build. Others are called at query time!
But yes, I will add it.